# Value proposition for new book app

You've been given a database of one of the services competing in the book app market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

In [1]:
#importing libraries
import pandas as pd
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'
pd.set_option('display.max_columns', 50)
from sqlalchemy import create_engine
#! pip install psycopg2
import psycopg2 as pc
#! pip install pymysql
#import pymysql

In [2]:
# connecting to db
db_config = {'user': 'practicum_student',         # username             
                      'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password         's65BlTKV3faNIGhmvJVzOqhs'  'Sdf4$2;d-d30pp'
                      'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',             
                      'port': 6432,              # connection port             
                      'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                                                             db_config['pwd'], 
                                                                                             db_config['host'], 
                                                                                             db_config['port'], 
                                                                                             db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def query_result(q):
    return pd.io.sql.read_sql(q, con = engine)

* **Find the number of books released after January 1, 2000.**

In [4]:
query = '''select *
from books
limit 3'''

query_result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [5]:
query = '''select count(book_id) as book_count
from books
where publication_date > '2000-01-01';'''

query_result(query)

book_count
0         819

the number of books released after 1/1/2000 is 819

* **Find the number of user reviews and the average rating for each book.**

In [6]:
query = '''select *
from reviews
limit 3'''

query_result(query)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

In [7]:
query = '''select *
from ratings
limit 3'''

query_result(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [10]:
query = '''select books.book_id, max(books.title) as title, count(reviews.review_id) as review_count, 
avg(ratings.rating) as avg_rating
from books
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
group by 1
order by 1 --3 desc;'''

query_result(query)

book_id                                              title  review_count  \
0          1                                       'Salem's Lot             6   
1          2                 1 000 Places to See Before You Die             2   
2          3  13 Little Blue Envelopes (Little Blue Envelope...             9   
3          4  1491: New Revelations of the Americas Before C...             4   
4          5                                               1776            24   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)             9   
996      997                        Xenocide (Ender's Saga  #3)            15   
997      998                                    Year of Wonders            20   
998      999                        You Suck (A Love Story  #2)             4   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...            24   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

Done

* **Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).**

In [9]:
query = '''select *
from books
limit 3;'''

query_result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [10]:
query = '''select *
from publishers
limit 3;'''

query_result(query)

publisher_id  publisher
0             1        Ace
1             2   Ace Book
2             3  Ace Books

In [11]:
query = '''select max(publishers.publisher) as publisher, count(books.book_id) as book_count
from books
LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
where books.num_pages > 50
group by books.publisher_id
order by 2 desc
limit 1;'''

query_result(query)

publisher  book_count
0  Penguin Books          42

the publisher that has released the greatest number of books with more than 50 pages is Penguin Books

* **Identify the author with the highest average book rating: look only at books with at least 50 ratings.**

In [12]:
query = '''select *
from books
limit 3;'''

query_result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135

In [13]:
query = '''select *
from ratings
limit 3'''

query_result(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [14]:
query = '''select *
from authors
limit 3'''

query_result(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie

In [15]:
query = '''SELECT max(authors.author) as author, books.book_id, max(authors.author_id) as author_id, 
count(ratings.rating_id) as rating_count, avg(ratings.rating) as avg_rating
FROM books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN authors ON authors.author_id = books.author_id
group by 2
having count(ratings.rating_id) > 50
order by 5 desc
limit 1;'''

query_result(query)

author  book_id  author_id  rating_count  avg_rating
0  J.K. Rowling/Mary GrandPré      302        236            82    4.414634

the author with the highest average book rating among books with at least 50 ratings is J.K. Rowling/Mary GrandPré

* **Find the average number of text reviews among users who rated more than 50 books.**

In [16]:
query = '''select *
from reviews
limit 3'''

query_result(query)

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   
2          3        2      lorichen   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...

In [17]:
query = '''select *
from ratings
limit 3'''

query_result(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5

In [18]:
query = '''
select avg(review_id) as avg_review_amount
from reviews
where username in
(select username
from ratings
group by 1
having count(rating_id) > 50);'''

query_result(query)

avg_review_amount
0        1315.760274

In [28]:
query = '''
select avg(count) as avg_review_amount
from
(select max(username) as username, count(review_id)
from reviews
where username in
(select username
from ratings
group by username
having count(rating_id) > 50)) as subq;'''

query_result(query)

avg_review_amount
0              146.0

In [29]:
query = '''
SELECT AVG(count) AS avg_review_number
FROM (SELECT username,
             COUNT (DISTINCT review_id)
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(DISTINCT book_id) > 50)
      GROUP BY username) s;'''

query_result(query)

avg_review_number
0          24.333333

the average amount of text reviews among users who rated more than 50 books is 24.333333